python/src/schema.rs

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use arrow::pyarrow::PyArrowType;
+use arrow_schema::Schema as ArrowSchema;
+use lance::datatypes::{Field, Schema};
+use lance_file::datatypes::Fields;
+use lance_file::format::pb;
+use prost::Message;
+use pyo3::{
+    basic::CompareOp,
+    exceptions::{PyNotImplementedError, PyValueError},
+    prelude::*,
+    types::PyTuple,
+};
+
+/// A Lance Schema.
+///
+/// This is valid for a particular dataset. It contains the field ids for each
+/// column in the dataset.
+#[pyclass(name = "LanceSchema", module = "lance.schema")]
+#[derive(Clone)]
+pub struct LanceSchema(pub Schema);
+
+#[pymethods]
+impl LanceSchema {
+    pub fn __repr__(&self) -> PyResult<String> {
+        // TODO: we should make a more succinct representation
+        Ok(format!("{:?}", self.0))
+    }
+
+    pub fn __richcmp__(&self, other: LanceSchema, op: CompareOp) -> PyResult<bool> {
+        match op {
+            CompareOp::Eq => Ok(self.0 == other.0),
+            CompareOp::Ne => Ok(self.0 != other.0),
+            _ => Err(PyNotImplementedError::new_err(
+                "Only == and != are supported",
+            )),
+        }
+    }
+
+    /// Convert the schema to a PyArrow schema.
+    pub fn to_pyarrow(&self) -> PyArrowType<ArrowSchema> {
+        PyArrowType(ArrowSchema::from(&self.0))
+    }
+
+    /// Create a Lance schema from a PyArrow schema.
+    ///
+    /// This will assign field ids in depth-first order. Be aware this may not
+    /// match the correct schema for a particular table.
+    #[staticmethod]
+    pub fn from_pyarrow(schema: PyArrowType<ArrowSchema>) -> PyResult<Self> {
+        let schema = Schema::try_from(&schema.0)
+            .map_err(|err| PyValueError::new_err(format!("Failed to convert schema: {}", err)))?;
+        Ok(Self(schema))
+    }
+
+    pub fn __reduce__(&self, py: Python<'_>) -> PyResult<(PyObject, PyObject)> {
+        // Each field is a protobuf message
+        let mut states = Vec::new();
+
+        let fields = Fields::from(&self.0);
+        for field in fields.0.iter() {
+            states.push(field.encode_to_vec());
+        }
+        // The metadata is its own JSON object
+        let metadata_str = serde_json::to_string(&self.0.metadata)
+            .map_err(|e| PyErr::new::<PyValueError, _>(format!("{}", e)))?;
+        states.push(metadata_str.as_bytes().to_vec());
+
+        let state = PyTuple::new(py, states).extract()?;
+        let from_json = PyModule::import(py, "lance.schema")?
+            .getattr("LanceSchema")?
+            .getattr("_from_protos")?
+            .extract()?;
+        Ok((from_json, state))
```

Comment:
```suggestion
        let from_protos = PyModule::import(py, "lance.schema")?
            .getattr("LanceSchema")?
            .getattr("_from_protos")?
            .extract()?;
        Ok((from_protos, state))
```

Although, arguably, `from_protos` isn't entirely accurate since the first arg is JSON but :shrug: Still, better to be consistent with the method name I think.